### ISSUE TITLE
Setting output_type of a cell to "error" in case of an error

### ISSUE THREAD

**Issue by AnotherCodeArtist** [gibiansky/IHaskell#837]
Hi!

I'd love to use IHaskell with nbgrader. Unfortunately nbgrader cannot identify failed tests in Haskell notebooks. After going through the sources of nbgrader and IHaskell the problem is the following:

nbgrader checks the cells in the notebook. If a cell is of `output_type=='error'` than it marks this cell as a failed test and does not assign points.
In all other cases it considers the cell as a passed test and grants full points.

IHaskell apparently sets the `output_type` to `"display_data"` regardless whether there was an error or not.

Could you please fix this?

**Comment by vaibhavsagar** [comment]
I've had a quick look at the source, and I can't see where we're setting `output_type` at all. If you know where we're doing this, can you point it out to me :smile:?

**Comment by AnotherCodeArtist** [comment]
That's going to happen here: https://github.com/gibiansky/IHaskell/blob/59c80f7af17a6e74ad525efe3331aaef97bcf220/src/IHaskell/Convert/LhsToIpynb.hs#L70

**Comment by vaibhavsagar** [comment]
I don't think so, that module is only used when converting `.lhs` files to `.ipynb` files.

**Comment by AnotherCodeArtist** [comment]
That's perfectly fine. `nbgrader evaluate` is first running the notebook and then inspecting it's resulting `ipynb` file. The only problem is figuring out, whether the current cell represents an erroneous result or not. I'd guess - although kind of a nasty hack - testing for the existence of the style class `err-msg` in the HTML text could work.

**Comment by AnotherCodeArtist** [comment]
Here' some more input:

When forcing an error in a Python notebook, the erroneous cell is represented in _ipynb_ as follows:

```json
{
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'wewee' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-2387f3e6a785>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mwewee\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'wewee' is not defined"
     ]
    }
   ],
   "source": [
    "wewee"
   ]
  }
``` 

Whereas in IHaskell it looks like this:

```json
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "cell-d4789f09616c0ccc",
     "locked": true,
     "points": 2,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>/* Styles used for the Hoogle display in the pager */\n",
       ".hoogle-doc {\n",
       "display: block;\n",
       "padding-bottom: 1.3em;\n",
       "padding-left: 0.4em;\n",
       "}\n",
       ".hoogle-code {\n",
       "display: block;\n",
       "font-family: monospace;\n",
       "white-space: pre;\n",
       "}\n",
       ".hoogle-text {\n",
       "display: block;\n",
       "}\n",
       ".hoogle-name {\n",
       "color: green;\n",
       "font-weight: bold;\n",
       "}\n",
       ".hoogle-head {\n",
       "font-weight: bold;\n",
       "}\n",
       ".hoogle-sub {\n",
       "display: block;\n",
       "margin-left: 0.4em;\n",
       "}\n",
       ".hoogle-package {\n",
       "font-weight: bold;\n",
       "font-style: italic;\n",
       "}\n",
       ".hoogle-module {\n",
       "font-weight: bold;\n",
       "}\n",
       ".hoogle-class {\n",
       "font-weight: bold;\n",
       "}\n",
       ".get-type {\n",
       "color: green;\n",
       "font-weight: bold;\n",
       "font-family: monospace;\n",
       "display: block;\n",
       "white-space: pre-wrap;\n",
       "}\n",
       ".show-type {\n",
       "color: green;\n",
       "font-weight: bold;\n",
       "font-family: monospace;\n",
       "margin-left: 1em;\n",
       "}\n",
       ".mono {\n",
       "font-family: monospace;\n",
       "display: block;\n",
       "}\n",
       ".err-msg {\n",
       "color: red;\n",
       "font-style: italic;\n",
       "font-family: monospace;\n",
       "white-space: pre;\n",
       "display: block;\n",
       "}\n",
       "#unshowable {\n",
       "color: red;\n",
       "font-weight: bold;\n",
       "}\n",
       ".err-msg.in.collapse {\n",
       "padding-top: 0.7em;\n",
       "}\n",
       ".highlight-code {\n",
       "white-space: pre;\n",
       "font-family: monospace;\n",
       "}\n",
       ".suggestion-warning { \n",
       "font-weight: bold;\n",
       "color: rgb(200, 130, 0);\n",
       "}\n",
       ".suggestion-error { \n",
       "font-weight: bold;\n",
       "color: red;\n",
       "}\n",
       ".suggestion-name {\n",
       "font-weight: bold;\n",
       "}\n",
       "</style><span class='err-msg'>&lt;interactive&gt;:1:1: error: Variable not in scope: zuizi :: IHaskell.Display.Height -&gt; t0 -&gt; t1 -&gt; t<br/>&lt;interactive&gt;:1:9: error:<br/>    • Variable not in scope: jhg<br/>    • Perhaps you meant ‘IHaskell.Display.jpg’ (imported from IHaskell.Display)<br/>&lt;interactive&gt;:1:13: error: Variable not in scope: j</span>"
      ],
      "text/plain": [
       "<interactive>:1:1: error: Variable not in scope: zuizi :: IHaskell.Display.Height -> t0 -> t1 -> t\n",
       "<interactive>:1:9: error:\n",
       "    • Variable not in scope: jhg\n",
       "    • Perhaps you meant ‘IHaskell.Display.jpg’ (imported from IHaskell.Display)\n",
       "<interactive>:1:13: error: Variable not in scope: j"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "import Control.Exception\n",
    "--assert (add 5 2 == 7) \"OK\"\n",
    "zuizi h jhg j"
   ]
  }
```

**Comment by vaibhavsagar** [comment]
My mistake, I think I found where we're setting `display_output`:

https://github.com/gibiansky/IHaskell/blob/33edf12f1328ef75bd3e347c5ba96f04dda1e752/src/IHaskell/Publish.hs#L65

I'll look into changing this so we correctly set `error`.

**Comment by SKogelnik** [comment]
It would be really nice to get IHaskell working with nbgrader. Is anyone working on a solution for this problem at the moment?

**Comment by vaibhavsagar** [comment]
I'm not aware of anyone currently working on this.

**Comment by vaibhavsagar** [comment]
I've started work on a PR for this at https://github.com/gibiansky/IHaskell/pull/1008.

**Comment by vaibhavsagar** [comment]
Closed by #1008.
